### PR TITLE
ci: align the dev-main-release promotion lifecycle

### LINKED ISSUE #345: [Feature]: Align CI lifecycle with the dev-main-release promotion flow
### Area
CI / release / workflows

### Problem statement
The repository's active branch model is `dev -> main -> release`, but the core CI workflows still behave like `main` is the only long-lived integration branch. As a result, `dev` push activity does not emit the expected CI status for the default branch, promotion rules stop at `dev -> main`, and contributor-facing docs drift away from the actual release hardening path.

### Proposed solution
Align `ci.yml`, `codeql.yml`, `security.yml`, and `perf-lint.yml` with the `dev -> main -> release` promotion chain, add a stable aggregate `build` check for branch protection, add a release-lane enforcement workflow so only `main` can merge into `release` branches, and update contributor/reliability docs to describe the same branch lifecycle.

### Non-goals / out of scope
Do not change the tag-driven `release.yml` publishing flow, do not auto-create release branches, and do not add broad bot automation outside the promotion guardrails.

### Alternatives considered
Keep CI PR-only or `main`-only and rely on manual promotion discipline, or require branch protection to track every internal CI job name individually. Both options leave `dev` without a stable default-branch status surface and keep repository governance brittle when CI jobs change.

### Acceptance criteria
- `CI`, `CodeQL`, and `Security` run for push and pull request activity on `dev`, `main`, `release`, and `release/*`
- `perf-lint` follows the same branch set while remaining path-scoped
- `CI` emits a stable aggregate `build` check for branch protection
- Promotion PRs into `main` still require `dev`, and promotion PRs into release lanes require `main`
- Contributor and reliability docs describe the same branch and CI lifecycle

### Policy / security / breaking sensitivity
Yes, this touches policy, security, or potentially breaking behavior

### Rollout / rollback notes
Land the workflow and documentation changes first, then switch branch protection to require the aggregate `build` check. If the new branch filters prove too broad, revert the workflow branch matrix and the release-lane enforcement workflow together so the documented branch model and automation stay in sync.

### Impact
Affected users: maintainers and contributors working against `dev`, `main`, and future release-hardening branches
Frequency: every push, promotion PR, and CI status check on the repository
Consequence if missing: the default branch keeps showing missing CI status, promotion guardrails stay asymmetric, and contributor docs remain out of sync with the actual release flow

### Submission checks
- [x] I searched existing issues and did not find the same request already tracked.
- [x] I used project-scoped wording and removed secrets or private data from examples.

### PR DESCRIPTION
**Pull request by chumyin**
## Summary

- Problem: `dev` is the primary development branch, but the repository workflows did not expose a stable CI status for the real `dev -> main -> release` promotion path, which left `dev` showing `No status` and left the release lane partially unenforced.
- Why it matters: branch protection and promotion governance only work when the required checks and merge guards match the repository's actual branch lifecycle.
- What changed: expanded the CI, CodeQL, security, and perf-lint workflow triggers to `dev`, `main`, `release`, and `release/**`; added an aggregate `build` job in `ci.yml` for a stable required check; added `enforce-main-to-release.yml`; aligned contributor and reliability docs with the promotion flow; added an implementation plan record.
- What did not change (scope boundary): this does not change Rust runtime behavior, crate boundaries, or release artifact semantics; it only aligns workflow triggering, promotion guards, and supporting docs.

## Linked Issues

- Closes #345
- Related #N/A

## Change Type

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [x] Documentation
- [ ] Security hardening
- [x] CI / workflow / release

## Touched Areas

- [ ] Kernel / policy / approvals
- [ ] Contracts / protocol / spec
- [ ] Daemon / CLI / install
- [ ] Providers / routing
- [ ] Tools
- [ ] Browser automation
- [ ] Channels / integrations
- [ ] ACP / conversation / session runtime
- [ ] Memory / context assembly
- [ ] Config / migration / onboarding
- [x] Docs / contributor workflow
- [x] CI / release / workflows

## Risk Track

- [ ] Track A (routine / low-risk)
- [x] Track B (higher-risk / policy-impacting)

If Track B, fill these in:

- Risk notes: workflow trigger changes affect which branches receive required checks and add a new enforcement rule on the `main -> release` lane.
- Rollout / guardrails: use the aggregate `build` job as the stable required check, and review the first `dev -> main` and `main -> release` PRs to confirm the branch filters and enforcement behavior match repository intent.
- Rollback path: revert this PR, or disable the new `enforce-main-to-release.yml` workflow if it blocks an urgent release while the policy is being adjusted.

## Validation

- [x] `cargo fmt --all -- --check`
- [x] `cargo clippy --workspace --all-targets --all-features -- -D warnings`
- [x] `cargo test --workspace --locked`
- [x] `cargo test --workspace --all-features --locked`
- [x] Relevant architecture / dep-graph / docs checks for touched areas
- [x] Additional scenario, benchmark, or manual checks when behavior changed
- [ ] If this changes config/env fallback, limits, or defaults: include before/after behavior and regression coverage for explicit path, fallback path, and boundary values
- [ ] If tests mutate process-global env: document how state is restored or serialized

Commands and evidence:

```text
cargo test --workspace --locked
cargo fmt --all -- --check
cargo clippy --workspace --all-targets --all-features -- -D warnings
cargo test --workspace --all-features --locked
./scripts/check_dep_graph.sh
LOONGCLAW_ARCH_STRICT=true ./scripts/check_architecture_boundaries.sh
ruby -e 'require "yaml"; ARGV.each { |path| YAML.load_file(path) }' .github/workflows/ci.yml .github/workflows/codeql.yml .github/workflows/security.yml .github/workflows/perf-lint.yml .github/workflows/enforce-main-to-release.yml .github/workflows/enforce-dev-to-main.yml
scripts/bootstrap_release_local_artifacts.sh
LOONGCLAW_RELEASE_DOCS_STRICT=1 scripts/check-docs.sh
git diff --check

All listed commands passed locally. The docs strict check required bootstrapping local `.docs` release artifacts first, which matches the repository's local governance flow.
```

## User-visible / Operator-visible Changes

- `dev` now receives CI runs on push and PR activity, with an aggregate `build` check that can be used as the stable required status.
- `main -> release` promotion is now explicitly enforced alongside the existing `dev -> main` guard.
- Contributor and reliability docs now describe the actual `dev -> main -> release` promotion lane and the checks that gate it.

## Failure Recovery

- Fast rollback or disable path: revert this PR, or temporarily disable `enforce-main-to-release.yml` if the policy needs to be relaxed during an incident.
- Observable failure symptoms reviewers should watch for: missing `build` status on `dev`-targeted PRs, unexpected `perf-lint` fan-out on unrelated pushes, or legitimate `main -> release` PRs being auto-closed incorrectly.

## Reviewer Focus

- Confirm the branch trigger matrix matches the intended `dev -> main -> release` lifecycle across `ci.yml`, `codeql.yml`, `security.yml`, and `perf-lint.yml`.
- Confirm the aggregate `build` job is the right stable status target for branch protection and fails whenever any required upstream job fails.
- Confirm `enforce-main-to-release.yml` mirrors the existing governance pattern cleanly and the docs describe the same policy the workflows enforce.

## Summary by CodeRabbit

* **Chores**
  * Extended CI workflows to run on dev, main, release, and release/** branches and broadened path triggers.
  * Added an aggregate "build" check that requires upstream CI jobs to succeed before completing.
  * Added enforcement that release-branch PRs must originate from main and strengthened dev→main source checks.

* **Tests**
  * Added scripted checks to validate promotion-guard workflow conditions.

* **Documentation**
  * Updated CONTRIBUTING and collaboration docs and added planning guidance for the new CI lifecycle.

### CODE REVIEW

**Comment by chumyin** [comment]
Follow-up commit `81bae550` addresses the promotion-guard same-repository bypass and the plan-doc review nits. All checks on the current head are green; the remaining merge blocker is the `dev` branch protection requirement for one approving review.

File: .github/workflows/enforce-main-to-release.yml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      - 'release/**'
+    types:
+      - opened
+      - reopened
+      - synchronize
+      - edited
+
+jobs:
+  block-non-main-source:
+    if: ${{ github.event.pull_request.head.ref != 'main' }}
```
[the remark this answers is not in the document]
Comment (chumyin) [reply .github/workflows/enforce-main-to-release.yml]: Fixed on the current head. The guard now requires both the same repository and the expected source branch, and the close message reports `owner/repo:branch` for the blocked source.

File: docs/plans/2026-03-18-dev-main-release-ci-lifecycle-implementation-plan.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+Keep the current `paths` filter, but add matching branch filters so:
+- `pull_request.branches` explicitly includes `dev`, `main`, `release`, and `release/**`
+- `push.branches` explicitly includes `dev`, `main`, `release`, and `release/**`
+- `push.paths` matches the existing path-scoped files so normal pushes do not trigger unrelated perf lint runs
+
+This keeps the benchmark-lint workflow scoped to relevant files while making it available on `dev`.
+
+**Step 5: Add release-lane promotion enforcement**
+
+Create `.github/workflows/enforce-main-to-release.yml` mirroring the style of
```
[the remark this answers is not in the document]
Comment (chumyin) [reply docs/plans/2026-03-18-dev-main-release-ci-lifecycle-implementation-plan.md:57]: Checked against the current file before changing anything. There was no `Github` or `GITHUB` occurrence in this document on the reviewed head, so no code change was needed for that item.

File: docs/plans/2026-03-18-dev-main-release-ci-lifecycle-implementation-plan.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+**Goal:** Align GitHub Actions with the repository's `dev -> main -> release` promotion model so `dev` receives continuous validation and branch protection can rely on a stable required check.
+
+**Architecture:** Keep the current workflow split between CI, security, CodeQL, perf lint, and release publishing, but change the branch filters so the integration workflows follow the promotion chain instead of only `main`. Add a final aggregate `build` job in `ci.yml` so branch protection can depend on one stable check name even if individual jobs change later.
+
+**Tech Stack:** GitHub Actions workflow YAML, shell validation, Rust workspace verification, Markdown planning docs
+
+---
+
+### Task 1: Align integration workflow trigger branches with the promotion chain
```
[the remark this answers is not in the document]
Comment (chumyin) [reply docs/plans/2026-03-18-dev-main-release-ci-lifecycle-implementation-plan.md]: Fixed on the current head. The task headings now use `##`, so the document no longer skips heading levels.
